hawkbit-repository/hawkbit-repository-core/src/main/java/org/eclipse/hawkbit/event/BusProtoStuffMessageConverter.java

```diff
@@ -0,0 +1,108 @@
+
+import io.protostuff.LinkedBuffer;
+import io.protostuff.ProtobufIOUtil;
+import io.protostuff.ProtostuffIOUtil;
+import io.protostuff.Schema;
+import io.protostuff.runtime.RuntimeSchema;
+
+/**
+ * A customize message converter for the spring cloud events. The converter is
+ * registered fot the application/binary+protostuff type.
+ *
+ */
+public class BusProtoStuffMessageConverter extends AbstractMessageConverter {
+
+    public static final MimeType APPLICATION_BINARY_PROTOSTUFF = new MimeType("application", "binary+protostuff");
+    private static final Logger LOG = LoggerFactory.getLogger(BusProtoStuffMessageConverter.class);
+    private static final String DEFAULT_CLASS_FIELD_NAME = "__Class__";
+
+    /**
+     * Constructor.
+     */
+    public BusProtoStuffMessageConverter() {
+        super(APPLICATION_BINARY_PROTOSTUFF);
+    }
+
+    @Override
+    protected boolean supports(final Class<?> aClass) {
+        return RemoteApplicationEvent.class.isAssignableFrom(aClass);
+    }
+
+    @Override
+    public Object convertFromInternal(final Message<?> message, final Class<?> targetClass,
+            final Object conversionHint) {
+        final Object payload = message.getPayload();
+
+        try {
+            final Class<?> eventClass = ClassUtils
+                    .getClass(message.getHeaders().get(DEFAULT_CLASS_FIELD_NAME).toString());
+            if (payload instanceof byte[]) {
+                @SuppressWarnings("unchecked")
+                final Schema<Object> schema = (Schema<Object>) RuntimeSchema.getSchema(eventClass);
+                final Object deserializeEvent = schema.newMessage();
+                ProtobufIOUtil.mergeFrom((byte[]) message.getPayload(), deserializeEvent, schema);
+                return deserializeEvent;
+            }
+        } catch (final ClassNotFoundException e) {
+            LOG.error("Protostuff cannot find derserialize class", e);
+            throw new MessageConversionException(message, "Failed to read payload", e);
+        }
+
+        return null;
+    }
+
+    @Override
+    protected Object convertToInternal(final Object payload, final MessageHeaders headers,
+            final Object conversionHint) {
+        checkIfHeaderMutable(headers);
+        final Class<? extends Object> eventClass = payload.getClass();
+        @SuppressWarnings("unchecked")
+        final Schema<Object> schema = (Schema<Object>) RuntimeSchema.getSchema(eventClass);
+        final LinkedBuffer buffer = LinkedBuffer.allocate();
+        final byte[] protostuff;
+        try {
+            protostuff = ProtostuffIOUtil.toByteArray(payload, schema, buffer);
+        } finally {
+            buffer.clear();
+        }
+
+        headers.put(DEFAULT_CLASS_FIELD_NAME, eventClass.getName());
+        return protostuff;
+    }
+
+    private static void checkIfHeaderMutable(final MessageHeaders headers) {
+        if (isAccessorMutable(headers) || headers instanceof MutableMessageHeaders) {
```

Comment:
One mutable header is a inner class, so yes you have to decide here. The abstract class do the same with the accessor, but i don't know who use the accessor. We will use the public ```MutableMessageHeaders```.